# Basic Exercises on Data Importing - Understanding - Manipulating - Analysis - Visualization

## Section-1: The pupose of the below exercises (1-7) is to create dictionary and convert into dataframes, how to diplay etc...
## The below exercises required to create data 

### 1. Import the necessary libraries (pandas, numpy, datetime, re etc)

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import datetime as dt
import seaborn as sns
import re

# set the graphs to show in the jupyter notebook
%matplotlib inline

# set seabor graphs to a better style
sns.set(style="ticks")

### 2. Run the below line of code to create a dictionary and this will be used for below exercises

In [3]:
raw_data = {"name": ['Bulbasaur', 'Charmander','Squirtle','Caterpie'],
            "evolution": ['Ivysaur','Charmeleon','Wartortle','Metapod'],
            "type": ['grass', 'fire', 'water', 'bug'],
            "hp": [45, 39, 44, 45],
            "pokedex": ['yes', 'no','yes','no']                        
            }

### 3. Assign it to a object called pokemon and it should be a pandas DataFrame

In [4]:
pokemon = pd.DataFrame(raw_data)
pokemon

name   evolution   type  hp pokedex
0   Bulbasaur     Ivysaur  grass  45     yes
1  Charmander  Charmeleon   fire  39      no
2    Squirtle   Wartortle  water  44     yes
3    Caterpie     Metapod    bug  45      no

### 4. If the DataFrame columns are in alphabetical order, change the order of the columns as name, type, hp, evolution, pokedex

In [5]:
pokemon = pokemon[['name', 'type', 'hp', 'evolution', 'pokedex']]
pokemon

name   type  hp   evolution pokedex
0   Bulbasaur  grass  45     Ivysaur     yes
1  Charmander   fire  39  Charmeleon      no
2    Squirtle  water  44   Wartortle     yes
3    Caterpie    bug  45     Metapod      no

### 5. Add another column called place, and insert places (lakes, parks, hills, forest etc) of your choice.

In [6]:
pokemon.insert(5, 'places', ['lakes', 'parks', 'hills', 'forest'])
pokemon

name   type  hp   evolution pokedex  places
0   Bulbasaur  grass  45     Ivysaur     yes   lakes
1  Charmander   fire  39  Charmeleon      no   parks
2    Squirtle  water  44   Wartortle     yes   hills
3    Caterpie    bug  45     Metapod      no  forest

### 6. Display the data type of each column

In [7]:
pokemon.dtypes

name         object
type         object
hp            int64
evolution    object
pokedex      object
places       object
dtype: object

### 7. Display the info of dataframe

In [8]:
pokemon.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 6 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   name       4 non-null      object
 1   type       4 non-null      object
 2   hp         4 non-null      int64 
 3   evolution  4 non-null      object
 4   pokedex    4 non-null      object
 5   places     4 non-null      object
dtypes: int64(1), object(5)
memory usage: 320.0+ bytes


## Section-2: The pupose of the below exercise (8-20) is to understand deleting data with pandas.
## The below exercises required to use wine.data

### 8. Import the dataset *wine.txt* from the folder and assign it to a object called wine

Please note that the original data text file doesn't contain any header. Please ensure that when you import the data, you should use a suitable argument so as to avoid data getting imported as header.

In [9]:
wine = pd.read_csv('Exercise Data Files\wine.txt', header = None)

In [10]:
wine.head()

0      1     2     3     4    5     6     7     8     9     10    11    12  \
0   1  14.23  1.71  2.43  15.6  127  2.80  3.06  0.28  2.29  5.64  1.04  3.92   
1   1  13.20  1.78  2.14  11.2  100  2.65  2.76  0.26  1.28  4.38  1.05  3.40   
2   1  13.16  2.36  2.67  18.6  101  2.80  3.24  0.30  2.81  5.68  1.03  3.17   
3   1  14.37  1.95  2.50  16.8  113  3.85  3.49  0.24  2.18  7.80  0.86  3.45   
4   1  13.24  2.59  2.87  21.0  118  2.80  2.69  0.39  1.82  4.32  1.04  2.93   

     13  
0  1065  
1  1050  
2  1185  
3  1480  
4   735

### 9. Delete the first, fourth, seventh, nineth, eleventh, thirteenth and fourteenth columns

In [11]:
wine.drop(columns = [0, 3, 6, 8, 10, 12, 13], inplace = True)

In [12]:
wine.head()

1     2     4    5     7     9     11
0  14.23  1.71  15.6  127  3.06  2.29  1.04
1  13.20  1.78  11.2  100  2.76  1.28  1.05
2  13.16  2.36  18.6  101  3.24  2.81  1.03
3  14.37  1.95  16.8  113  3.49  2.18  0.86
4  13.24  2.59  21.0  118  2.69  1.82  1.04

### 10. Assign the columns as below:

The attributes are (dontated by Riccardo Leardi, riclea '@' anchem.unige.it):  
1) alcohol  
2) malic_acid  
3) alcalinity_of_ash  
4) magnesium  
5) flavanoids  
6) proanthocyanins  
7) hue 

In [13]:
wine.columns = ['alcohol', 'malic_acid', 'alcalinity_of_ash', 'magnesium', 'flavanoids', 'proanthocyanins', 'hue']

In [14]:
wine.head()

alcohol  malic_acid  alcalinity_of_ash  magnesium  flavanoids  \
0    14.23        1.71               15.6        127        3.06   
1    13.20        1.78               11.2        100        2.76   
2    13.16        2.36               18.6        101        3.24   
3    14.37        1.95               16.8        113        3.49   
4    13.24        2.59               21.0        118        2.69   

   proanthocyanins   hue  
0             2.29  1.04  
1             1.28  1.05  
2             2.81  1.03  
3             2.18  0.86  
4             1.82  1.04

### 11. Set the values of the first 3 values from alcohol column as NaN

In [15]:
wine.loc[0:2, 'alcohol'] = np.nan

In [16]:
wine.head()

alcohol  malic_acid  alcalinity_of_ash  magnesium  flavanoids  \
0      NaN        1.71               15.6        127        3.06   
1      NaN        1.78               11.2        100        2.76   
2      NaN        2.36               18.6        101        3.24   
3    14.37        1.95               16.8        113        3.49   
4    13.24        2.59               21.0        118        2.69   

   proanthocyanins   hue  
0             2.29  1.04  
1             1.28  1.05  
2             2.81  1.03  
3             2.18  0.86  
4             1.82  1.04

### 12. Now set the value of the rows 3 and 4 of magnesium as NaN

In [17]:
wine.loc[2:3, 'magnesium'] = np.nan
wine.head()

alcohol  malic_acid  alcalinity_of_ash  magnesium  flavanoids  \
0      NaN        1.71               15.6      127.0        3.06   
1      NaN        1.78               11.2      100.0        2.76   
2      NaN        2.36               18.6        NaN        3.24   
3    14.37        1.95               16.8        NaN        3.49   
4    13.24        2.59               21.0      118.0        2.69   

   proanthocyanins   hue  
0             2.29  1.04  
1             1.28  1.05  
2             2.81  1.03  
3             2.18  0.86  
4             1.82  1.04

### 13. Fill the value of NaN with the number 10 in alcohol and 100 in magnesium

In [18]:
wine = wine.fillna(value = {'alcohol': 10, 'magnesium': 100})
wine.head()

alcohol  malic_acid  alcalinity_of_ash  magnesium  flavanoids  \
0    10.00        1.71               15.6      127.0        3.06   
1    10.00        1.78               11.2      100.0        2.76   
2    10.00        2.36               18.6      100.0        3.24   
3    14.37        1.95               16.8      100.0        3.49   
4    13.24        2.59               21.0      118.0        2.69   

   proanthocyanins   hue  
0             2.29  1.04  
1             1.28  1.05  
2             2.81  1.03  
3             2.18  0.86  
4             1.82  1.04

### 14. Count the number of missing values in all columns.

In [19]:
wine.isna().sum()

alcohol              0
malic_acid           0
alcalinity_of_ash    0
magnesium            0
flavanoids           0
proanthocyanins      0
hue                  0
dtype: int64

### 15.  Create an array of 10 random numbers up until 10 and save it.

In [20]:
random_arr = np.random.randint(10, size = 10)
random_arr

array([1, 5, 4, 7, 0, 2, 1, 7, 5, 4])

### 16.  Set the rows corresponding to the random numbers to NaN in the column *alcohol*

In [21]:
wine.loc[random_arr, 'alcohol'] = np.nan
wine.head(10)

alcohol  malic_acid  alcalinity_of_ash  magnesium  flavanoids  \
0      NaN        1.71               15.6      127.0        3.06   
1      NaN        1.78               11.2      100.0        2.76   
2      NaN        2.36               18.6      100.0        3.24   
3    14.37        1.95               16.8      100.0        3.49   
4      NaN        2.59               21.0      118.0        2.69   
5      NaN        1.76               15.2      112.0        3.39   
6    14.39        1.87               14.6       96.0        2.52   
7      NaN        2.15               17.6      121.0        2.51   
8    14.83        1.64               14.0       97.0        2.98   
9    13.86        1.35               16.0       98.0        3.15   

   proanthocyanins   hue  
0             2.29  1.04  
1             1.28  1.05  
2             2.81  1.03  
3             2.18  0.86  
4             1.82  1.04  
5             1.97  1.05  
6             1.98  1.02  
7             1.25  1.06  
8             1.98  1.08  
9             1.85  1.01

### 17.  How many missing values do we have now?

In [22]:
wine.isna().sum()

alcohol              6
malic_acid           0
alcalinity_of_ash    0
magnesium            0
flavanoids           0
proanthocyanins      0
hue                  0
dtype: int64

### 18. Print only the non-null values in alcohol

In [23]:
wine.loc[wine['alcohol'].notnull()]

alcohol  malic_acid  alcalinity_of_ash  magnesium  flavanoids  \
3      14.37        1.95               16.8      100.0        3.49   
6      14.39        1.87               14.6       96.0        2.52   
8      14.83        1.64               14.0       97.0        2.98   
9      13.86        1.35               16.0       98.0        3.15   
10     14.10        2.16               18.0      105.0        3.32   
..       ...         ...                ...        ...         ...   
173    13.71        5.65               20.5       95.0        0.61   
174    13.40        3.91               23.0      102.0        0.75   
175    13.27        4.28               20.0      120.0        0.69   
176    13.17        2.59               20.0      120.0        0.68   
177    14.13        4.10               24.5       96.0        0.76   

     proanthocyanins   hue  
3               2.18  0.86  
6               1.98  1.02  
8               1.98  1.08  
9               1.85  1.01  
10              2.38  1.25  
..               ...   ...  
173             1.06  0.64  
174             1.41  0.70  
175             1.35  0.59  
176             1.46  0.60  
177             1.35  0.61  

[172 rows x 7 columns]

### 19. Delete the rows that contain missing values

In [24]:
wine.dropna(inplace= True)

In [25]:
wine

alcohol  malic_acid  alcalinity_of_ash  magnesium  flavanoids  \
3      14.37        1.95               16.8      100.0        3.49   
6      14.39        1.87               14.6       96.0        2.52   
8      14.83        1.64               14.0       97.0        2.98   
9      13.86        1.35               16.0       98.0        3.15   
10     14.10        2.16               18.0      105.0        3.32   
..       ...         ...                ...        ...         ...   
173    13.71        5.65               20.5       95.0        0.61   
174    13.40        3.91               23.0      102.0        0.75   
175    13.27        4.28               20.0      120.0        0.69   
176    13.17        2.59               20.0      120.0        0.68   
177    14.13        4.10               24.5       96.0        0.76   

     proanthocyanins   hue  
3               2.18  0.86  
6               1.98  1.02  
8               1.98  1.08  
9               1.85  1.01  
10              2.38  1.25  
..               ...   ...  
173             1.06  0.64  
174             1.41  0.70  
175             1.35  0.59  
176             1.46  0.60  
177             1.35  0.61  

[172 rows x 7 columns]

### 20.  Reset the index, so it starts with 0 again

In [26]:
wine.reset_index(drop = True, inplace = True)

In [27]:
wine

alcohol  malic_acid  alcalinity_of_ash  magnesium  flavanoids  \
0      14.37        1.95               16.8      100.0        3.49   
1      14.39        1.87               14.6       96.0        2.52   
2      14.83        1.64               14.0       97.0        2.98   
3      13.86        1.35               16.0       98.0        3.15   
4      14.10        2.16               18.0      105.0        3.32   
..       ...         ...                ...        ...         ...   
167    13.71        5.65               20.5       95.0        0.61   
168    13.40        3.91               23.0      102.0        0.75   
169    13.27        4.28               20.0      120.0        0.69   
170    13.17        2.59               20.0      120.0        0.68   
171    14.13        4.10               24.5       96.0        0.76   

     proanthocyanins   hue  
0               2.18  0.86  
1               1.98  1.02  
2               1.98  1.08  
3               1.85  1.01  
4               2.38  1.25  
..               ...   ...  
167             1.06  0.64  
168             1.41  0.70  
169             1.35  0.59  
170             1.46  0.60  
171             1.35  0.61  

[172 rows x 7 columns]

## Section-3: The pupose of the below exercise (21-27) is to understand ***filtering & sorting*** data from dataframe.
## The below exercises required to use chipotle.tsv

This time we are going to pull data directly from the internet.  
Import the dataset directly from this link (https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv) and create dataframe called chipo

In [28]:
chipo = pd.read_csv('Exercise Data Files\chipotle.tsv', sep = '\t')
chipo

order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
1            1         1                                   Izze   
2            1         1                       Nantucket Nectar   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
4            2         2                           Chicken Bowl   
...        ...       ...                                    ...   
4617      1833         1                          Steak Burrito   
4618      1833         1                          Steak Burrito   
4619      1834         1                     Chicken Salad Bowl   
4620      1834         1                     Chicken Salad Bowl   
4621      1834         1                     Chicken Salad Bowl   

                                     choice_description item_price  
0                                                   NaN     $2.39   
1                                          [Clementine]     $3.39   
2                                               [Apple]     $3.39   
3                                                   NaN     $2.39   
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98   
...                                                 ...        ...  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...    $11.75   
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...    $11.75   
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25   
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75   
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75   

[4622 rows x 5 columns]

### 21. How many products cost more than $10.00? 

Use `str` attribute to remove the $ sign and convert the column to proper numeric type data before filtering.


In [29]:
chipo['item_price'] = chipo['item_price'].str.replace('$','', regex = False)
chipo['item_price'] = pd.to_numeric(chipo['item_price'])

In [30]:
chipo[chipo['item_price']>10]['item_name'].count()

1130

### 22. Print the Chipo Dataframe & info about data frame

In [31]:
print(chipo)

      order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
1            1         1                                   Izze   
2            1         1                       Nantucket Nectar   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
4            2         2                           Chicken Bowl   
...        ...       ...                                    ...   
4617      1833         1                          Steak Burrito   
4618      1833         1                          Steak Burrito   
4619      1834         1                     Chicken Salad Bowl   
4620      1834         1                     Chicken Salad Bowl   
4621      1834         1                     Chicken Salad Bowl   

                                     choice_description  item_price  
0                                                   NaN        2.39  
1                                          [Clementine]

In [32]:
print(chipo.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   order_id            4622 non-null   int64  
 1   quantity            4622 non-null   int64  
 2   item_name           4622 non-null   object 
 3   choice_description  3376 non-null   object 
 4   item_price          4622 non-null   float64
dtypes: float64(1), int64(2), object(2)
memory usage: 180.7+ KB
None


### 23. What is the price of each item? 
- Delete the duplicates in item_name and quantity
- Print a data frame with only two columns `item_name` and `item_price`
- Sort the values from the most to less expensive

In [33]:
chipo_ = chipo.copy()
chipo_.drop_duplicates(subset = ['item_name', 'quantity'], inplace = True, ignore_index = True)

In [34]:
chipo_ = chipo[['item_name', 'item_price']]
chipo_

item_name  item_price
0              Chips and Fresh Tomato Salsa        2.39
1                                      Izze        3.39
2                          Nantucket Nectar        3.39
3     Chips and Tomatillo-Green Chili Salsa        2.39
4                              Chicken Bowl       16.98
...                                     ...         ...
4617                          Steak Burrito       11.75
4618                          Steak Burrito       11.75
4619                     Chicken Salad Bowl       11.25
4620                     Chicken Salad Bowl        8.75
4621                     Chicken Salad Bowl        8.75

[4622 rows x 2 columns]

In [35]:
chipo_.sort_values(by = 'item_price', ascending = False)

item_name  item_price
3598  Chips and Fresh Tomato Salsa       44.25
3480                 Carnitas Bowl       35.25
1254               Chicken Burrito       35.00
3602               Chicken Burrito       35.00
3601                Veggie Burrito       33.75
...                            ...         ...
3936                   Canned Soda        1.09
2922                 Bottled Water        1.09
1396                   Canned Soda        1.09
2562                   Canned Soda        1.09
1457                   Canned Soda        1.09

[4622 rows x 2 columns]

### 24. Sort by the name of the item

In [36]:
chipo.sort_values(by = 'item_name', inplace = True)
chipo

order_id  quantity          item_name  \
3389      1360         2  6 Pack Soft Drink   
341        148         1  6 Pack Soft Drink   
1849       749         1  6 Pack Soft Drink   
1860       754         1  6 Pack Soft Drink   
2713      1076         1  6 Pack Soft Drink   
...        ...       ...                ...   
2384       948         1  Veggie Soft Tacos   
781        322         1  Veggie Soft Tacos   
2851      1132         1  Veggie Soft Tacos   
1699       688         1  Veggie Soft Tacos   
1395       567         1  Veggie Soft Tacos   

                                     choice_description  item_price  
3389                                        [Diet Coke]       12.98  
341                                         [Diet Coke]        6.49  
1849                                             [Coke]        6.49  
1860                                        [Diet Coke]        6.49  
2713                                             [Coke]        6.49  
...                                                 ...         ...  
2384  [Roasted Chili Corn Salsa, [Fajita Vegetables,...        8.75  
781   [Fresh Tomato Salsa, [Black Beans, Cheese, Sou...        8.75  
2851  [Roasted Chili Corn Salsa (Medium), [Black Bea...        8.49  
1699  [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...       11.25  
1395  [Fresh Tomato Salsa (Mild), [Pinto Beans, Rice...        8.49  

[4622 rows x 5 columns]

### 25. What was the quantity of the most expensive item ordered?

In [37]:
chipo[chipo['item_price'] == chipo['item_price'].max()]['quantity'].sum()

15

### 26. How many times were a Veggie Salad Bowl ordered?

In [38]:
chipo[chipo['item_name'] == 'Veggie Salad Bowl']['item_name'].count()

18

### 27. How many times people orderd more than one Canned Soda?

In [39]:
chipo[(chipo['item_name'] == 'Canned Soda') & (chipo['quantity'] > 1)]['item_name'].count()

20

## Section-4: The purpose of the below exercises is to understand how to perform aggregations of data frame
## The below exercises (28-33) required to use occupation.csv

###  28. Import the dataset occupation.csv and assign object as users

In [40]:
users = pd.read_csv('Exercise Data Files\occupation.csv', sep = '|')
users

user_id  age gender     occupation zip_code
0          1   24      M     technician    85711
1          2   53      F          other    94043
2          3   23      M         writer    32067
3          4   24      M     technician    43537
4          5   33      F          other    15213
..       ...  ...    ...            ...      ...
938      939   26      F        student    33319
939      940   32      M  administrator    02215
940      941   20      M        student    97229
941      942   48      F      librarian    78209
942      943   22      M        student    77841

[943 rows x 5 columns]

### 29. Discover what is the mean age per occupation

In [41]:
users.groupby('occupation').age.mean().reset_index()

occupation        age
0   administrator  38.746835
1          artist  31.392857
2          doctor  43.571429
3        educator  42.010526
4        engineer  36.388060
5   entertainment  29.222222
6       executive  38.718750
7      healthcare  41.562500
8       homemaker  32.571429
9          lawyer  36.750000
10      librarian  40.000000
11      marketing  37.615385
12           none  26.555556
13          other  34.523810
14     programmer  33.121212
15        retired  63.071429
16       salesman  35.666667
17      scientist  35.548387
18        student  22.081633
19     technician  33.148148
20         writer  36.311111

### 30. Discover the Male ratio per occupation and sort it from the most to the least.

Use numpy.where() to encode gender column.

In [42]:
users['is_male'] = np.where(users.gender == 'M', 1, 0)
users

user_id  age gender     occupation zip_code  is_male
0          1   24      M     technician    85711        1
1          2   53      F          other    94043        0
2          3   23      M         writer    32067        1
3          4   24      M     technician    43537        1
4          5   33      F          other    15213        0
..       ...  ...    ...            ...      ...      ...
938      939   26      F        student    33319        0
939      940   32      M  administrator    02215        1
940      941   20      M        student    97229        1
941      942   48      F      librarian    78209        0
942      943   22      M        student    77841        1

[943 rows x 6 columns]

In [43]:
(users.groupby('occupation').is_male.sum()/users.groupby('occupation').gender.count()).round(2).sort_values(ascending = False)

occupation
doctor           1.00
engineer         0.97
technician       0.96
retired          0.93
executive        0.91
programmer       0.91
scientist        0.90
entertainment    0.89
lawyer           0.83
salesman         0.75
educator         0.73
student          0.69
other            0.66
marketing        0.62
writer           0.58
none             0.56
administrator    0.54
artist           0.54
librarian        0.43
healthcare       0.31
homemaker        0.14
dtype: float64

### 31. For each occupation, calculate the minimum and maximum ages

In [44]:
users.groupby('occupation').age.agg(['min', 'max'])

min  max
occupation             
administrator   21   70
artist          19   48
doctor          28   64
educator        23   63
engineer        22   70
entertainment   15   50
executive       22   69
healthcare      22   62
homemaker       20   50
lawyer          21   53
librarian       23   69
marketing       24   55
none            11   55
other           13   64
programmer      20   63
retired         51   73
salesman        18   66
scientist       23   55
student          7   42
technician      21   55
writer          18   60

### 32. For each combination of occupation and gender, calculate the mean age

In [45]:
users.groupby(['occupation', 'gender']).age.mean()

occupation     gender
administrator  F         40.638889
               M         37.162791
artist         F         30.307692
               M         32.333333
doctor         M         43.571429
educator       F         39.115385
               M         43.101449
engineer       F         29.500000
               M         36.600000
entertainment  F         31.000000
               M         29.000000
executive      F         44.000000
               M         38.172414
healthcare     F         39.818182
               M         45.400000
homemaker      F         34.166667
               M         23.000000
lawyer         F         39.500000
               M         36.200000
librarian      F         40.000000
               M         40.000000
marketing      F         37.200000
               M         37.875000
none           F         36.500000
               M         18.600000
other          F         35.472222
               M         34.028986
programmer     F         32.16666

### 33.  For each occupation present the percentage of women and men

In [46]:
(users.groupby('occupation').gender.value_counts(normalize = True)*100).round(2)

occupation     gender
administrator  M          54.43
               F          45.57
artist         M          53.57
               F          46.43
doctor         M         100.00
educator       M          72.63
               F          27.37
engineer       M          97.01
               F           2.99
entertainment  M          88.89
               F          11.11
executive      M          90.62
               F           9.38
healthcare     F          68.75
               M          31.25
homemaker      F          85.71
               M          14.29
lawyer         M          83.33
               F          16.67
librarian      F          56.86
               M          43.14
marketing      M          61.54
               F          38.46
none           M          55.56
               F          44.44
other          M          65.71
               F          34.29
programmer     M          90.91
               F           9.09
retired        M          92.86
               F  

## Section-6: The purpose of the below exercises is to understand how to use lambda-apply-functions
## The below exercises (34-41) required to use student-mat.csv and student-por.csv files 

### 34. Import the datasets *student-mat* and *student-por* and append them and assigned object as df

In [47]:
df = pd.concat((pd.read_csv('Exercise Data Files\student-mat.csv'), pd.read_csv('Exercise Data Files\student-por.csv')))
df

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0       GP   F   18       U     GT3       A     4     4   at_home   teacher   
1       GP   F   17       U     GT3       T     1     1   at_home     other   
2       GP   F   15       U     LE3       T     1     1   at_home     other   
3       GP   F   15       U     GT3       T     4     2    health  services   
4       GP   F   16       U     GT3       T     3     3     other     other   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
644     MS   F   19       R     GT3       T     2     3  services     other   
645     MS   F   18       U     LE3       T     3     1   teacher  services   
646     MS   F   18       U     GT3       T     1     1     other     other   
647     MS   M   17       U     LE3       T     3     1  services  services   
648     MS   M   18       R     LE3       T     3     2  services     other   

     ... famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0    ...      4        3      4     1     1      3        6   5   6   6  
1    ...      5        3      3     1     1      3        4   5   5   6  
2    ...      4        3      2     2     3      3       10   7   8  10  
3    ...      3        2      2     1     1      5        2  15  14  15  
4    ...      4        3      2     1     2      5        4   6  10  10  
..   ...    ...      ...    ...   ...   ...    ...      ...  ..  ..  ..  
644  ...      5        4      2     1     2      5        4  10  11  10  
645  ...      4        3      4     1     1      1        4  15  15  16  
646  ...      1        1      1     1     1      5        6  11  12   9  
647  ...      2        4      5     3     4      2        6  10  10  10  
648  ...      4        4      1     3     4      5        4  10  11  11  

[1044 rows x 33 columns]

### 35. For the purpose of this exercise slice the dataframe from 'school' until the 'guardian' column

In [48]:
df = df.loc[:, 'school':'guardian']
df

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0       GP   F   18       U     GT3       A     4     4   at_home   teacher   
1       GP   F   17       U     GT3       T     1     1   at_home     other   
2       GP   F   15       U     LE3       T     1     1   at_home     other   
3       GP   F   15       U     GT3       T     4     2    health  services   
4       GP   F   16       U     GT3       T     3     3     other     other   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
644     MS   F   19       R     GT3       T     2     3  services     other   
645     MS   F   18       U     LE3       T     3     1   teacher  services   
646     MS   F   18       U     GT3       T     1     1     other     other   
647     MS   M   17       U     LE3       T     3     1  services  services   
648     MS   M   18       R     LE3       T     3     2  services     other   

     reason guardian  
0    course   mother  
1    course   father  
2     other   mother  
3      home   mother  
4      home   father  
..      ...      ...  
644  course   mother  
645  course   mother  
646  course   mother  
647  course   mother  
648  course   mother  

[1044 rows x 12 columns]

### 36. Create a lambda function that captalize strings (example: if we give at_home as input function and should give At_home as output.

In [49]:
Cap = lambda x: x.capitalize()
print(Cap('at_home'))

At_home


### 37. Capitalize both Mjob and Fjob variables using above lamdba function

In [50]:
df['Mjob'].apply(Cap)
df['Fjob'].apply(Cap)

0       Teacher
1         Other
2         Other
3      Services
4         Other
         ...   
644       Other
645    Services
646       Other
647    Services
648       Other
Name: Fjob, Length: 1044, dtype: object

### 38. Print the last elements of the data set. (Last few records)

In [51]:
df.tail()

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
644     MS   F   19       R     GT3       T     2     3  services     other   
645     MS   F   18       U     LE3       T     3     1   teacher  services   
646     MS   F   18       U     GT3       T     1     1     other     other   
647     MS   M   17       U     LE3       T     3     1  services  services   
648     MS   M   18       R     LE3       T     3     2  services     other   

     reason guardian  
644  course   mother  
645  course   mother  
646  course   mother  
647  course   mother  
648  course   mother

### 39. Did you notice the original dataframe is still lowercase? Why is that? Fix it and captalize Mjob and Fjob.

In [52]:
df['Mjob'] = df['Mjob'].apply(Cap)
df['Fjob'] = df['Fjob'].apply(Cap)
df

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0       GP   F   18       U     GT3       A     4     4   At_home   Teacher   
1       GP   F   17       U     GT3       T     1     1   At_home     Other   
2       GP   F   15       U     LE3       T     1     1   At_home     Other   
3       GP   F   15       U     GT3       T     4     2    Health  Services   
4       GP   F   16       U     GT3       T     3     3     Other     Other   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
644     MS   F   19       R     GT3       T     2     3  Services     Other   
645     MS   F   18       U     LE3       T     3     1   Teacher  Services   
646     MS   F   18       U     GT3       T     1     1     Other     Other   
647     MS   M   17       U     LE3       T     3     1  Services  Services   
648     MS   M   18       R     LE3       T     3     2  Services     Other   

     reason guardian  
0    course   mother  
1    course   father  
2     other   mother  
3      home   mother  
4      home   father  
..      ...      ...  
644  course   mother  
645  course   mother  
646  course   mother  
647  course   mother  
648  course   mother  

[1044 rows x 12 columns]

### 40. Create a function called majority that return a boolean value to a new column called legal_drinker

In [53]:
def majority(n):
    return n >= 18

df['legal_drinker'] = majority(df['age'])
df

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0       GP   F   18       U     GT3       A     4     4   At_home   Teacher   
1       GP   F   17       U     GT3       T     1     1   At_home     Other   
2       GP   F   15       U     LE3       T     1     1   At_home     Other   
3       GP   F   15       U     GT3       T     4     2    Health  Services   
4       GP   F   16       U     GT3       T     3     3     Other     Other   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
644     MS   F   19       R     GT3       T     2     3  Services     Other   
645     MS   F   18       U     LE3       T     3     1   Teacher  Services   
646     MS   F   18       U     GT3       T     1     1     Other     Other   
647     MS   M   17       U     LE3       T     3     1  Services  Services   
648     MS   M   18       R     LE3       T     3     2  Services     Other   

     reason guardian  legal_drinker  
0    course   mother           True  
1    course   father          False  
2     other   mother          False  
3      home   mother          False  
4      home   father          False  
..      ...      ...            ...  
644  course   mother           True  
645  course   mother           True  
646  course   mother           True  
647  course   mother          False  
648  course   mother           True  

[1044 rows x 13 columns]

### 41. Multiply every number of the dataset by 10. 

In [54]:
df[df.select_dtypes(include = 'number').columns] *= 10
df

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0       GP   F  180       U     GT3       A    40    40   At_home   Teacher   
1       GP   F  170       U     GT3       T    10    10   At_home     Other   
2       GP   F  150       U     LE3       T    10    10   At_home     Other   
3       GP   F  150       U     GT3       T    40    20    Health  Services   
4       GP   F  160       U     GT3       T    30    30     Other     Other   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
644     MS   F  190       R     GT3       T    20    30  Services     Other   
645     MS   F  180       U     LE3       T    30    10   Teacher  Services   
646     MS   F  180       U     GT3       T    10    10     Other     Other   
647     MS   M  170       U     LE3       T    30    10  Services  Services   
648     MS   M  180       R     LE3       T    30    20  Services     Other   

     reason guardian  legal_drinker  
0    course   mother           True  
1    course   father          False  
2     other   mother          False  
3      home   mother          False  
4      home   father          False  
..      ...      ...            ...  
644  course   mother           True  
645  course   mother           True  
646  course   mother           True  
647  course   mother          False  
648  course   mother           True  

[1044 rows x 13 columns]

## Section-6: The purpose of the below exercises is to understand how to perform simple joins
## The below exercises (42-48) required to use cars1.csv and cars2.csv files 

### 42. Import the datasets cars1.csv and cars2.csv and assign names as cars1 and cars2

In [55]:
cars1 = pd.read_csv('Exercise Data Files\cars1.csv')
cars1

mpg  cylinders  displacement horsepower  weight  acceleration  model  \
0    18.0          8           307        130    3504          12.0     70   
1    15.0          8           350        165    3693          11.5     70   
2    18.0          8           318        150    3436          11.0     70   
3    16.0          8           304        150    3433          12.0     70   
4    17.0          8           302        140    3449          10.5     70   
..    ...        ...           ...        ...     ...           ...    ...   
193  24.0          6           200         81    3012          17.6     76   
194  22.5          6           232         90    3085          17.6     76   
195  29.0          4            85         52    2035          22.2     76   
196  24.5          4            98         60    2164          22.1     76   
197  29.0          4            90         70    1937          14.2     76   

     origin                        car  Unnamed: 9  Unnamed: 10  Unnamed: 11  \
0         1  chevrolet chevelle malibu         NaN          NaN          NaN   
1         1          buick skylark 320         NaN          NaN          NaN   
2         1         plymouth satellite         NaN          NaN          NaN   
3         1              amc rebel sst         NaN          NaN          NaN   
4         1                ford torino         NaN          NaN          NaN   
..      ...                        ...         ...          ...          ...   
193       1              ford maverick         NaN          NaN          NaN   
194       1                 amc hornet         NaN          NaN          NaN   
195       1         chevrolet chevette         NaN          NaN          NaN   
196       1            chevrolet woody         NaN          NaN          NaN   
197       2                  vw rabbit         NaN          NaN          NaN   

     Unnamed: 12  Unnamed: 13  
0            NaN          NaN  
1            NaN          NaN  
2            NaN          NaN  
3            NaN          NaN  
4            NaN          NaN  
..           ...          ...  
193          NaN          NaN  
194          NaN          NaN  
195          NaN          NaN  
196          NaN          NaN  
197          NaN          NaN  

[198 rows x 14 columns]

In [56]:
cars2 = pd.read_csv('Exercise Data Files\cars2.csv')
cars2

mpg  cylinders  displacement horsepower  weight  acceleration  model  \
0    33.0          4            91         53    1795          17.4     76   
1    20.0          6           225        100    3651          17.7     76   
2    18.0          6           250         78    3574          21.0     76   
3    18.5          6           250        110    3645          16.2     76   
4    17.5          6           258         95    3193          17.8     76   
..    ...        ...           ...        ...     ...           ...    ...   
195  27.0          4           140         86    2790          15.6     82   
196  44.0          4            97         52    2130          24.6     82   
197  32.0          4           135         84    2295          11.6     82   
198  28.0          4           120         79    2625          18.6     82   
199  31.0          4           119         82    2720          19.4     82   

     origin                 car  
0         3         honda civic  
1         1      dodge aspen se  
2         1   ford granada ghia  
3         1  pontiac ventura sj  
4         1       amc pacer d/l  
..      ...                 ...  
195       1     ford mustang gl  
196       2           vw pickup  
197       1       dodge rampage  
198       1         ford ranger  
199       1          chevy s-10  

[200 rows x 9 columns]

   ### 43. Print the information to cars1 by applying below functions 
   hint: Use different functions/methods like type(), head(), tail(), columns(), info(), dtypes(), index(), shape(), count(), size(), ndim(), axes(), describe(), memory_usage(), sort_values(), value_counts()
   Also create profile report using pandas_profiling.Profile_Report

In [57]:
type(cars1)

pandas.core.frame.DataFrame

In [58]:
cars1.head()

mpg  cylinders  displacement horsepower  weight  acceleration  model  \
0  18.0          8           307        130    3504          12.0     70   
1  15.0          8           350        165    3693          11.5     70   
2  18.0          8           318        150    3436          11.0     70   
3  16.0          8           304        150    3433          12.0     70   
4  17.0          8           302        140    3449          10.5     70   

   origin                        car  Unnamed: 9  Unnamed: 10  Unnamed: 11  \
0       1  chevrolet chevelle malibu         NaN          NaN          NaN   
1       1          buick skylark 320         NaN          NaN          NaN   
2       1         plymouth satellite         NaN          NaN          NaN   
3       1              amc rebel sst         NaN          NaN          NaN   
4       1                ford torino         NaN          NaN          NaN   

   Unnamed: 12  Unnamed: 13  
0          NaN          NaN  
1          NaN          NaN  
2          NaN          NaN  
3          NaN          NaN  
4          NaN          NaN

In [59]:
cars1.tail()

mpg  cylinders  displacement horsepower  weight  acceleration  model  \
193  24.0          6           200         81    3012          17.6     76   
194  22.5          6           232         90    3085          17.6     76   
195  29.0          4            85         52    2035          22.2     76   
196  24.5          4            98         60    2164          22.1     76   
197  29.0          4            90         70    1937          14.2     76   

     origin                 car  Unnamed: 9  Unnamed: 10  Unnamed: 11  \
193       1       ford maverick         NaN          NaN          NaN   
194       1          amc hornet         NaN          NaN          NaN   
195       1  chevrolet chevette         NaN          NaN          NaN   
196       1     chevrolet woody         NaN          NaN          NaN   
197       2           vw rabbit         NaN          NaN          NaN   

     Unnamed: 12  Unnamed: 13  
193          NaN          NaN  
194          NaN          NaN  
195          NaN          NaN  
196          NaN          NaN  
197          NaN          NaN

In [60]:
cars1.columns

Index(['mpg', 'cylinders', 'displacement', 'horsepower', 'weight',
       'acceleration', 'model', 'origin', 'car', 'Unnamed: 9', 'Unnamed: 10',
       'Unnamed: 11', 'Unnamed: 12', 'Unnamed: 13'],
      dtype='object')

In [61]:
cars1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 198 entries, 0 to 197
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           198 non-null    float64
 1   cylinders     198 non-null    int64  
 2   displacement  198 non-null    int64  
 3   horsepower    198 non-null    object 
 4   weight        198 non-null    int64  
 5   acceleration  198 non-null    float64
 6   model         198 non-null    int64  
 7   origin        198 non-null    int64  
 8   car           198 non-null    object 
 9   Unnamed: 9    0 non-null      float64
 10  Unnamed: 10   0 non-null      float64
 11  Unnamed: 11   0 non-null      float64
 12  Unnamed: 12   0 non-null      float64
 13  Unnamed: 13   0 non-null      float64
dtypes: float64(7), int64(5), object(2)
memory usage: 21.8+ KB


In [62]:
cars1.dtypes

mpg             float64
cylinders         int64
displacement      int64
horsepower       object
weight            int64
acceleration    float64
model             int64
origin            int64
car              object
Unnamed: 9      float64
Unnamed: 10     float64
Unnamed: 11     float64
Unnamed: 12     float64
Unnamed: 13     float64
dtype: object

In [63]:
cars1.index

RangeIndex(start=0, stop=198, step=1)

In [64]:
cars1.shape

(198, 14)

In [65]:
cars1.count()

mpg             198
cylinders       198
displacement    198
horsepower      198
weight          198
acceleration    198
model           198
origin          198
car             198
Unnamed: 9        0
Unnamed: 10       0
Unnamed: 11       0
Unnamed: 12       0
Unnamed: 13       0
dtype: int64

In [66]:
cars1.size

2772

In [67]:
cars1.ndim

2

In [68]:
cars1.axes

[RangeIndex(start=0, stop=198, step=1),
 Index(['mpg', 'cylinders', 'displacement', 'horsepower', 'weight',
        'acceleration', 'model', 'origin', 'car', 'Unnamed: 9', 'Unnamed: 10',
        'Unnamed: 11', 'Unnamed: 12', 'Unnamed: 13'],
       dtype='object')]

In [69]:
cars1.describe()

mpg   cylinders  displacement       weight  acceleration  \
count  198.000000  198.000000    198.000000   198.000000    198.000000   
mean    19.719697    5.898990    223.469697  3177.888889     15.005556   
std      5.814254    1.785417    115.181017   934.783733      2.872382   
min      9.000000    3.000000     68.000000  1613.000000      8.000000   
25%     15.000000    4.000000    113.250000  2302.500000     13.000000   
50%     19.000000    6.000000    228.000000  3030.000000     15.000000   
75%     24.375000    8.000000    318.000000  4080.750000     16.800000   
max     35.000000    8.000000    455.000000  5140.000000     23.500000   

            model      origin  Unnamed: 9  Unnamed: 10  Unnamed: 11  \
count  198.000000  198.000000         0.0          0.0          0.0   
mean    72.818182    1.439394         NaN          NaN          NaN   
std      1.865332    0.708085         NaN          NaN          NaN   
min     70.000000    1.000000         NaN          NaN          NaN   
25%     71.000000    1.000000         NaN          NaN          NaN   
50%     73.000000    1.000000         NaN          NaN          NaN   
75%     74.000000    2.000000         NaN          NaN          NaN   
max     76.000000    3.000000         NaN          NaN          NaN   

       Unnamed: 12  Unnamed: 13  
count          0.0          0.0  
mean           NaN          NaN  
std            NaN          NaN  
min            NaN          NaN  
25%            NaN          NaN  
50%            NaN          NaN  
75%            NaN          NaN  
max            NaN          NaN

In [70]:
cars1.memory_usage()

Index            128
mpg             1584
cylinders       1584
displacement    1584
horsepower      1584
weight          1584
acceleration    1584
model           1584
origin          1584
car             1584
Unnamed: 9      1584
Unnamed: 10     1584
Unnamed: 11     1584
Unnamed: 12     1584
Unnamed: 13     1584
dtype: int64

In [71]:
cars1.sort_values(by = 'mpg', ignore_index = True)

mpg  cylinders  displacement horsepower  weight  acceleration  model  \
0     9.0          8           304        193    4732          18.5     70   
1    10.0          8           307        200    4376          15.0     70   
2    10.0          8           360        215    4615          14.0     70   
3    11.0          8           400        150    4997          14.0     73   
4    11.0          8           429        208    4633          11.0     72   
..    ...        ...           ...        ...     ...           ...    ...   
193  31.0          4            76         52    1649          16.5     74   
194  32.0          4            83         61    2003          19.0     74   
195  32.0          4            71         65    1836          21.0     74   
196  33.0          4            91         53    1795          17.5     75   
197  35.0          4            72         69    1613          18.0     71   

     origin                  car  Unnamed: 9  Unnamed: 10  Unnamed: 11  \
0         1             hi 1200d         NaN          NaN          NaN   
1         1            chevy c20         NaN          NaN          NaN   
2         1            ford f250         NaN          NaN          NaN   
3         1     chevrolet impala         NaN          NaN          NaN   
4         1      mercury marquis         NaN          NaN          NaN   
..      ...                  ...         ...          ...          ...   
193       3        toyota corona         NaN          NaN          NaN   
194       3           datsun 710         NaN          NaN          NaN   
195       3  toyota corolla 1200         NaN          NaN          NaN   
196       3     honda civic cvcc         NaN          NaN          NaN   
197       3          datsun 1200         NaN          NaN          NaN   

     Unnamed: 12  Unnamed: 13  
0            NaN          NaN  
1            NaN          NaN  
2            NaN          NaN  
3            NaN          NaN  
4            NaN          NaN  
..           ...          ...  
193          NaN          NaN  
194          NaN          NaN  
195          NaN          NaN  
196          NaN          NaN  
197          NaN          NaN  

[198 rows x 14 columns]

In [72]:
cars1.value_counts()

Series([], dtype: int64)

In [73]:
#from pandas_profiling import ProfileReport
#ProfileReport(cars1)

### 44. It seems our first dataset has some unnamed blank columns, fix cars1

In [74]:
cars1 = cars1.loc[:, ~cars1.columns.str.contains('Unnamed')]
cars1

mpg  cylinders  displacement horsepower  weight  acceleration  model  \
0    18.0          8           307        130    3504          12.0     70   
1    15.0          8           350        165    3693          11.5     70   
2    18.0          8           318        150    3436          11.0     70   
3    16.0          8           304        150    3433          12.0     70   
4    17.0          8           302        140    3449          10.5     70   
..    ...        ...           ...        ...     ...           ...    ...   
193  24.0          6           200         81    3012          17.6     76   
194  22.5          6           232         90    3085          17.6     76   
195  29.0          4            85         52    2035          22.2     76   
196  24.5          4            98         60    2164          22.1     76   
197  29.0          4            90         70    1937          14.2     76   

     origin                        car  
0         1  chevrolet chevelle malibu  
1         1          buick skylark 320  
2         1         plymouth satellite  
3         1              amc rebel sst  
4         1                ford torino  
..      ...                        ...  
193       1              ford maverick  
194       1                 amc hornet  
195       1         chevrolet chevette  
196       1            chevrolet woody  
197       2                  vw rabbit  

[198 rows x 9 columns]

### 45. What is the number of observations in each dataset?

In [75]:
len(cars1)

198

In [76]:
len(cars2)

200

### 46. Join cars1 and cars2 into a single DataFrame called cars

In [77]:
cars = cars1.merge(cars2, how = 'outer')
cars

mpg  cylinders  displacement horsepower  weight  acceleration  model  \
0    18.0          8           307        130    3504          12.0     70   
1    15.0          8           350        165    3693          11.5     70   
2    18.0          8           318        150    3436          11.0     70   
3    16.0          8           304        150    3433          12.0     70   
4    17.0          8           302        140    3449          10.5     70   
..    ...        ...           ...        ...     ...           ...    ...   
393  27.0          4           140         86    2790          15.6     82   
394  44.0          4            97         52    2130          24.6     82   
395  32.0          4           135         84    2295          11.6     82   
396  28.0          4           120         79    2625          18.6     82   
397  31.0          4           119         82    2720          19.4     82   

     origin                        car  
0         1  chevrolet chevelle malibu  
1         1          buick skylark 320  
2         1         plymouth satellite  
3         1              amc rebel sst  
4         1                ford torino  
..      ...                        ...  
393       1            ford mustang gl  
394       2                  vw pickup  
395       1              dodge rampage  
396       1                ford ranger  
397       1                 chevy s-10  

[398 rows x 9 columns]

### 47. There is a column missing, called owners. Create a random number Series from 15,000 to 73,000.

In [78]:
pd.Series(np.random.randint(15000, 73000, size = len(cars)))

0      60412
1      56280
2      63070
3      30954
4      38271
       ...  
393    43419
394    44454
395    26247
396    28373
397    20533
Length: 398, dtype: int32

### 48. Add the column owners to cars

In [79]:
cars['owners'] = pd.Series(np.random.randint(15000, 73000, size = len(cars)))
cars

mpg  cylinders  displacement horsepower  weight  acceleration  model  \
0    18.0          8           307        130    3504          12.0     70   
1    15.0          8           350        165    3693          11.5     70   
2    18.0          8           318        150    3436          11.0     70   
3    16.0          8           304        150    3433          12.0     70   
4    17.0          8           302        140    3449          10.5     70   
..    ...        ...           ...        ...     ...           ...    ...   
393  27.0          4           140         86    2790          15.6     82   
394  44.0          4            97         52    2130          24.6     82   
395  32.0          4           135         84    2295          11.6     82   
396  28.0          4           120         79    2625          18.6     82   
397  31.0          4           119         82    2720          19.4     82   

     origin                        car  owners  
0         1  chevrolet chevelle malibu   40885  
1         1          buick skylark 320   17462  
2         1         plymouth satellite   26278  
3         1              amc rebel sst   67216  
4         1                ford torino   68662  
..      ...                        ...     ...  
393       1            ford mustang gl   66076  
394       2                  vw pickup   17651  
395       1              dodge rampage   60431  
396       1                ford ranger   54770  
397       1                 chevy s-10   63282  

[398 rows x 10 columns]

## Section-7: The purpose of the below exercises is to understand how to perform date time operations

### 49. Write a Python script to display the
- a. Current date and time
- b. Current year
- c. Month of year
- d. Week number of the year
- e. Weekday of the week
- f. Day of year
- g. Day of the month
- h. Day of week

In [80]:
# (a) Current date and time

print(dt.datetime.now())

2022-12-14 17:59:08.554125


In [81]:
# (b) Current year

print(dt.date.today().year)

2022


In [82]:
# (c) Month of year

print(dt.date.today().month)

12


In [83]:
# (d) Week number of the year

print(dt.date.today().strftime('%V'))

50


In [84]:
# (e) Weekday of the week

dt.date.today().strftime('%A')

'Wednesday'

In [85]:
# (f) Day of year

print(dt.date.today().strftime('%j'))

348


In [86]:
# (g) Day of the month

print(dt.date.today().strftime('%d'))

14


In [87]:
# (h) Day of week

dt.datetime.today().weekday()

2

### 50. Write a Python program to convert a string to datetime.
Sample String : Jul 1 2014 2:43PM 

Expected Output : 2014-07-01 14:43:00

In [88]:
def func1(s):
    output = pd.to_datetime(s, format = '%b %d %Y %I:%M%p')
    return output

func1('Jul 1 2014 2:43PM')

Timestamp('2014-07-01 14:43:00')

### 51. Write a Python program to subtract five days from current date.

Current Date : 2015-06-22

5 days before Current Date : 2015-06-17

In [89]:
def func2():
    output = dt.date.today() + pd.DateOffset(days = -5)
    return output.strftime("%Y-%m-%d")

func2()

'2022-12-09'

### 52. Write a Python program to convert unix timestamp string to readable date.

Sample Unix timestamp string : 1284105682
    
Expected Output : 2010-09-10 13:31:22

In [90]:
def func3(ts):
    return dt.datetime.utcfromtimestamp(ts).strftime('%Y-%m-%d %H:%M:%S')
   
func3(1284105682)

'2010-09-10 08:01:22'

### 53. Convert the below Series to pandas datetime : 

DoB = pd.Series(["07Sep59","01Jan55","15Dec47","11Jul42"])

Make sure that the year is 19XX not 20XX

In [91]:
DoB = pd.Series(["07Sep59","01Jan55","15Dec47","11Jul42"])
DoB = pd.to_datetime(DoB, format = '%d%b%y') + pd.DateOffset(days = -100*365.25)
DoB

0   1959-09-07
1   1955-01-01
2   1947-12-15
3   1942-07-11
dtype: datetime64[ns]

### 54. Write a Python program to get days between two dates. 

In [92]:
def func4(date1, date2):
    return (dt.datetime.strptime(date1, '%d-%m-%Y')- dt.datetime.strptime(date2, '%d-%m-%Y')).days

func4('21-07-2020', '21-07-2019')

366

### 55. Convert the below date to datetime and then change its display format using the .dt module

Date = "15Dec1989"

Result : "Friday, 15 Dec 98"

In [93]:
Date = "15Dec1998"
Date = dt.datetime.strptime(Date, '%d%b%Y')
Date.strftime('%A, %d %b %y')

'Tuesday, 15 Dec 98'

## The below exercises (56-66) required to use wind.data file 

### About wind.data:

The data have been modified to contain some missing values, identified by NaN.  

1. The data in 'wind.data' has the following format:

### 56. Import the dataset wind.data and assign it to a variable called data and replace the first 3 columns by a proper date time index

In [94]:
data = pd.read_csv('Exercise Data Files\wind.data',
                  parse_dates = {'dates' : ['Yr', 'Mo', 'Dy']},
                  keep_date_col=True)
data

dates  Yr  Mo  Dy    RPT    VAL    ROS    KIL    SHA    BIR    DUB  \
0    2061-01-01  61   1   1  15.04  14.96  13.17   9.29    NaN   9.87  13.67   
1    2061-01-02  61   1   2  14.71    NaN  10.83   6.50  12.62   7.67  11.50   
2    2061-01-03  61   1   3  18.50  16.88  12.33  10.13  11.17   6.17  11.25   
3    2061-01-04  61   1   4  10.58   6.63  11.75   4.58   4.54   2.88   8.63   
4    2061-01-05  61   1   5  13.33  13.25  11.42   6.17  10.71   8.21  11.92   
...         ...  ..  ..  ..    ...    ...    ...    ...    ...    ...    ...   
6569 1978-12-27  78  12  27  17.58  16.96  17.62   8.08  13.21  11.67  14.46   
6570 1978-12-28  78  12  28  13.21   5.46  13.46   5.00   8.12   9.42  14.33   
6571 1978-12-29  78  12  29  14.00  10.29  14.42   8.71   9.71  10.54  19.17   
6572 1978-12-30  78  12  30  18.50  14.04  21.29   9.13  12.75   9.71  18.08   
6573 1978-12-31  78  12  31  20.33  17.41  27.29   9.59  12.08  10.13  19.25   

        CLA    MUL    CLO    BEL    MAL  
0     10.25  10.83  12.58  18.50  15.04  
1     10.04   9.79   9.67  17.54  13.83  
2       NaN   8.50   7.67  12.75  12.71  
3      1.79   5.83   5.88   5.46  10.88  
4      6.54  10.92  10.34  12.92  11.83  
...     ...    ...    ...    ...    ...  
6569  15.59  14.04  14.00  17.21  40.08  
6570  16.25  15.25  18.05  21.79  41.46  
6571  12.46  14.50  16.42  18.88  29.58  
6572  12.87  12.46  12.12  14.67  28.79  
6573  11.63  11.58  11.38  12.08  22.08  

[6574 rows x 16 columns]

### 57. Year 2061 is seemingly imporoper. Convert every year which are < 70 to 19XX instead of 20XX.

In [95]:
data.loc[data['Yr'].astype(int) < 70, 'dates'] = data['dates'] + pd.DateOffset(days = -100 *365.25)
data

dates  Yr  Mo  Dy    RPT    VAL    ROS    KIL    SHA    BIR    DUB  \
0    1961-01-01  61   1   1  15.04  14.96  13.17   9.29    NaN   9.87  13.67   
1    1961-01-02  61   1   2  14.71    NaN  10.83   6.50  12.62   7.67  11.50   
2    1961-01-03  61   1   3  18.50  16.88  12.33  10.13  11.17   6.17  11.25   
3    1961-01-04  61   1   4  10.58   6.63  11.75   4.58   4.54   2.88   8.63   
4    1961-01-05  61   1   5  13.33  13.25  11.42   6.17  10.71   8.21  11.92   
...         ...  ..  ..  ..    ...    ...    ...    ...    ...    ...    ...   
6569 1978-12-27  78  12  27  17.58  16.96  17.62   8.08  13.21  11.67  14.46   
6570 1978-12-28  78  12  28  13.21   5.46  13.46   5.00   8.12   9.42  14.33   
6571 1978-12-29  78  12  29  14.00  10.29  14.42   8.71   9.71  10.54  19.17   
6572 1978-12-30  78  12  30  18.50  14.04  21.29   9.13  12.75   9.71  18.08   
6573 1978-12-31  78  12  31  20.33  17.41  27.29   9.59  12.08  10.13  19.25   

        CLA    MUL    CLO    BEL    MAL  
0     10.25  10.83  12.58  18.50  15.04  
1     10.04   9.79   9.67  17.54  13.83  
2       NaN   8.50   7.67  12.75  12.71  
3      1.79   5.83   5.88   5.46  10.88  
4      6.54  10.92  10.34  12.92  11.83  
...     ...    ...    ...    ...    ...  
6569  15.59  14.04  14.00  17.21  40.08  
6570  16.25  15.25  18.05  21.79  41.46  
6571  12.46  14.50  16.42  18.88  29.58  
6572  12.87  12.46  12.12  14.67  28.79  
6573  11.63  11.58  11.38  12.08  22.08  

[6574 rows x 16 columns]

In [96]:
data.drop(columns = ['Yr', 'Mo', 'Dy'], inplace = True)

### 58. Set the right dates as the index. Pay attention at the data type, it should be datetime64[ns].

In [97]:
data = data.set_index('dates')
data

RPT    VAL    ROS    KIL    SHA    BIR    DUB    CLA    MUL  \
dates                                                                       
1961-01-01  15.04  14.96  13.17   9.29    NaN   9.87  13.67  10.25  10.83   
1961-01-02  14.71    NaN  10.83   6.50  12.62   7.67  11.50  10.04   9.79   
1961-01-03  18.50  16.88  12.33  10.13  11.17   6.17  11.25    NaN   8.50   
1961-01-04  10.58   6.63  11.75   4.58   4.54   2.88   8.63   1.79   5.83   
1961-01-05  13.33  13.25  11.42   6.17  10.71   8.21  11.92   6.54  10.92   
...           ...    ...    ...    ...    ...    ...    ...    ...    ...   
1978-12-27  17.58  16.96  17.62   8.08  13.21  11.67  14.46  15.59  14.04   
1978-12-28  13.21   5.46  13.46   5.00   8.12   9.42  14.33  16.25  15.25   
1978-12-29  14.00  10.29  14.42   8.71   9.71  10.54  19.17  12.46  14.50   
1978-12-30  18.50  14.04  21.29   9.13  12.75   9.71  18.08  12.87  12.46   
1978-12-31  20.33  17.41  27.29   9.59  12.08  10.13  19.25  11.63  11.58   

              CLO    BEL    MAL  
dates                            
1961-01-01  12.58  18.50  15.04  
1961-01-02   9.67  17.54  13.83  
1961-01-03   7.67  12.75  12.71  
1961-01-04   5.88   5.46  10.88  
1961-01-05  10.34  12.92  11.83  
...           ...    ...    ...  
1978-12-27  14.00  17.21  40.08  
1978-12-28  18.05  21.79  41.46  
1978-12-29  16.42  18.88  29.58  
1978-12-30  12.12  14.67  28.79  
1978-12-31  11.38  12.08  22.08  

[6574 rows x 12 columns]

In [98]:
data.index

DatetimeIndex(['1961-01-01', '1961-01-02', '1961-01-03', '1961-01-04',
               '1961-01-05', '1961-01-06', '1961-01-07', '1961-01-08',
               '1961-01-09', '1961-01-10',
               ...
               '1978-12-22', '1978-12-23', '1978-12-24', '1978-12-25',
               '1978-12-26', '1978-12-27', '1978-12-28', '1978-12-29',
               '1978-12-30', '1978-12-31'],
              dtype='datetime64[ns]', name='dates', length=6574, freq=None)

### 59. Compute how many values are missing for each location over the entire record.  
#### They should be ignored in all calculations below. 

In [99]:
data.isna().sum()

RPT    6
VAL    3
ROS    2
KIL    5
SHA    2
BIR    0
DUB    3
CLA    2
MUL    3
CLO    1
BEL    0
MAL    4
dtype: int64

### 60. Compute how many non-missing values there are in total.

In [100]:
data.notna().sum().sum()

78857

### 61. Calculate the mean windspeeds over all the locations and all the times.
#### A single number for the entire dataset.

In [101]:
data.mean().mean()

10.227982360836924

### 62. Create a DataFrame called loc_stats and calculate the min, max and mean windspeeds and standard deviations of the windspeeds at each location over all the days 

#### A different set of numbers for each location.

In [102]:
loc_stats = pd.DataFrame(columns = data.columns)

In [103]:
loc_stats.loc['min', :] = data.min()
loc_stats.loc['max', :] = data.max()
loc_stats.loc['mean', :] = data.mean()
loc_stats.loc['std', :] = data.std()
loc_stats

RPT        VAL        ROS       KIL        SHA       BIR  \
min        0.67       0.21        1.5       0.0       0.13       0.0   
max        35.8      33.37      33.84     28.46      37.54     26.16   
mean  12.362987  10.644314  11.660526  6.306468  10.455834  7.092254   
std    5.618413   5.267356    5.00845  3.605811   4.936125  3.968683   

           DUB       CLA       MUL       CLO        BEL        MAL  
min        0.0       0.0       0.0      0.04       0.13       0.67  
max      30.37     31.08     25.88     28.21      42.38      42.54  
mean  9.797343  8.495053   8.49359  8.707332  13.121007  15.599079  
std   4.977555  4.499449  4.166872  4.503954   5.835037   6.699794

### 63. Create a DataFrame called day_stats and calculate the min, max and mean windspeed and standard deviations of the windspeeds across all the locations at each day.

#### A different set of numbers for each day.

In [104]:
day_stats = pd.DataFrame(index = data.index)

In [105]:
day_stats['min'] = data.min(axis = 1)
day_stats['max'] = data.max(axis = 1)
day_stats['mean'] = data.mean(axis = 1)
day_stats['std'] = data.std(axis = 1)

In [106]:
day_stats

min    max       mean       std
dates                                       
1961-01-01  9.29  18.50  13.018182  2.808875
1961-01-02  6.50  17.54  11.336364  3.188994
1961-01-03  6.17  18.50  11.641818  3.681912
1961-01-04  1.79  11.75   6.619167  3.198126
1961-01-05  6.17  13.33  10.630000  2.445356
...          ...    ...        ...       ...
1978-12-27  8.08  40.08  16.708333  7.868076
1978-12-28  5.00  41.46  15.150000  9.687857
1978-12-29  8.71  29.58  14.890000  5.756836
1978-12-30  9.13  28.79  15.367500  5.540437
1978-12-31  9.59  27.29  15.402500  5.702483

[6574 rows x 4 columns]

### 64. Find the average windspeed in January for each location.  
#### Treat January 1961 and January 1962 both as January.

In [107]:
data.loc[data.index.month == 1].mean()

RPT    14.847325
VAL    12.914560
ROS    13.299624
KIL     7.199498
SHA    11.667734
BIR     8.054839
DUB    11.819355
CLA     9.512047
MUL     9.543208
CLO    10.053566
BEL    14.550520
MAL    18.028763
dtype: float64

### 65. Calculate the mean windspeed for each month in the dataset.  
#### Treat January 1961 and January 1962 as *different* months.
#### (hint: first find a  way to create an identifier unique for each month.)

In [108]:
data_ = data.copy()
data_['year_month'] = data_.index.strftime('%Y-%m')
data_.groupby('year_month').mean()

RPT        VAL        ROS       KIL        SHA        BIR  \
year_month                                                                    
1961-01     14.841333  11.988333  13.431613  7.736774  11.072759   8.588065   
1961-02     16.269286  14.975357  14.441481  9.230741  13.852143  10.937500   
1961-03     10.890000  11.296452  10.752903  7.284000  10.509355   8.866774   
1961-04     10.722667   9.427667   9.998000  5.830667   8.435000   6.495000   
1961-05      9.860968   8.850000  10.818065  5.905333   9.490323   6.574839   
...               ...        ...        ...       ...        ...        ...   
2071-08      9.783226   6.991935  10.227097  4.540968   8.978065   4.797742   
2071-09      8.365000   6.512667   7.704333  3.470667   7.676667   5.462667   
2071-10     12.926129  11.106452  12.467419  5.778065  11.305161   8.562903   
2071-11     13.334667  10.853667  12.223333  6.250667  10.895000   7.660000   
2071-12     13.162581  10.606774  13.367097  6.070000  10.393871   7.841613   

                  DUB        CLA        MUL        CLO        BEL        MAL  
year_month                                                                    
1961-01     11.184839   9.245333   9.085806  10.107419  13.880968  14.703226  
1961-02     11.890714  11.846071  11.821429  12.714286  18.583214  15.411786  
1961-03      9.644194   9.829677  10.294138  11.251935  16.410968  15.720000  
1961-04      6.925333   7.094667   7.342333   7.237000  11.147333  10.278333  
1961-05      7.604000   8.177097   8.039355   8.499355  11.900323  12.011613  
...               ...        ...        ...        ...        ...        ...  
2071-08      7.450645   5.460645   6.866452   5.244194  10.087742  11.500968  
2071-09      6.003000   4.437333   5.385667   4.748667   9.232333  12.311333  
2071-10     10.073226   8.091935   9.121935   9.571290  15.617097  17.890645  
2071-11     11.062333   7.127667   9.714000   8.815000  15.145000  21.020333  
2071-12     12.131290   8.605161   9.708387  10.186129  16.759032  21.169355  

[216 rows x 12 columns]

In [109]:
data_ = data.copy()
data_["yearmonth"] = data_.index.strftime("%Y-%m")
data_.groupby(["yearmonth"]).apply(lambda x: pd.Series({"mean_windspeed": x.mean(numeric_only=True).mean()}))

mean_windspeed
yearmonth                
1961-01         11.322206
1961-02         13.497834
1961-03         11.062533
1961-04          8.411167
1961-05          8.977606
...                   ...
2071-08          7.660806
2071-09          6.775861
2071-10         11.042688
2071-11         11.175139
2071-12         11.666774

[216 rows x 1 columns]

### 66. Calculate the min, max and mean windspeeds and standard deviations of the windspeeds across all locations for each week (assume that the first week starts on January 2 1961) for the first 52 weeks.

In [110]:
weekly = data.resample('W').agg(['min', 'max', 'mean', 'std'])
weekly.loc[weekly.index[1:53],:]

RPT                                VAL                    \
              min    max       mean       std    min    max       mean   
dates                                                                    
1961-01-08  10.58  18.50  13.541429  2.631321   6.63  16.88  11.486667   
1961-01-15   9.04  19.75  12.468571  3.555392   3.54  12.08   8.967143   
1961-01-22   4.92  19.83  13.204286  5.337402   3.42  14.37   9.862857   
1961-01-29  13.62  25.04  19.880000  4.619061   9.96  23.91  16.141429   
1961-02-05  10.58  24.21  16.827143  5.251408   9.46  24.21  15.460000   
1961-02-12  16.00  24.54  19.684286  3.587677  11.54  21.42  16.417143   
1961-02-19   6.04  22.50  15.130000  5.064609  11.63  20.17  15.091429   
1961-02-26   7.79  25.80  15.221429  7.020716   7.08  21.50  13.625714   
1961-03-05  10.96  13.33  12.101429  0.997721   8.83  17.00  12.951429   
1961-03-12   4.88  14.79   9.376667  3.732263   8.08  16.96  11.578571   
1961-03-19   4.92  16.88  11.911429  3.860036   9.46  15.54  13.501429   
1961-03-26   6.29  15.00   9.567143  3.613298   2.58  11.63   8.387143   
1961-04-02   5.88  18.25  10.757143  5.046922   3.50  16.29   8.852857   
1961-04-09   4.50  18.12  11.964286  4.604392   7.04  14.62  10.654286   
1961-04-16   4.71  15.50   8.965714  3.937727   4.83  12.25   8.000000   
1961-04-23   4.00  21.09  12.621429  5.676655   3.71  15.41  10.438571   
1961-04-30   4.08  16.29  10.117143  4.349662   6.50  14.46   9.798571   
1961-05-07   9.87  23.00  15.367143  5.025507  10.29  19.79  13.970000   
1961-05-14   3.54  12.79   7.772857  3.371022   3.96  15.12   8.712857   
1961-05-21   4.88  15.04   8.225714  3.631730   3.58  10.17   5.631667   
1961-05-28   4.96  11.79   8.155714  2.739433   3.67  12.50   7.388571   
1961-06-04   7.00  15.92  10.321429  3.099701   4.75   9.79   7.407143   
1961-06-11   8.29  14.42  10.917143  2.248597   6.83  11.54   8.992857   
1961-06-18   6.13  14.33  10.571429  3.009482   4.12  14.54   9.565714   
1961-06-25   4.00   9.00   7.345714  1.982035   3.25   9.50   6.108571   
1961-07-02   7.21  13.13  10.236667  2.557856   6.34  14.37   9.482857   
1961-07-09   7.29  17.50  11.715714  3.664855   3.75  10.75   7.220000   
1961-07-16   8.63  22.50  16.680000  5.168710   7.87  19.29  13.518571   
1961-07-23   3.04   5.88   4.202857  1.047978   2.92   6.79   4.255714   
1961-07-30   6.13  16.08  10.561429  4.157641   4.63  13.79   8.445714   
1961-08-06   7.67  16.08  10.870000  2.950887   5.09  15.79   8.792857   
1961-08-13   2.88  14.21  10.058333  4.422268   4.42  10.00   7.941429   
1961-08-20  13.13  18.91  15.607143  2.283635   7.25  13.62  10.565714   
1961-08-27   7.67  18.16  12.391429  3.395857   6.87  14.58  11.430000   
1961-09-03   3.63  14.46   8.678571  4.398615   1.13  22.00   8.821429   
1961-09-10   5.00  17.62  10.541429  5.207278   3.04  13.59   8.798571   
1961-09-17   5.33  28.75  17.160000  7.679190   7.12  22.08  14.440000   
1961-09-24   6.92  10.25   8.500000  1.267399   2.92  13.62   7.154286   
1961-10-01  10.34  23.21  16.044286  4.559572   6.96  13.62  11.361667   
1961-10-08   3.13  16.08  11.250000  5.596710   3.63  16.96   8.757143   
1961-10-15   3.71  17.12  10.881667  4.780675   6.13  13.25   9.734286   
1961-10-22  10.46  28.62  19.260000  7.888314   3.75  19.46  13.364286   
1961-10-29   4.33  26.42  16.077143  7.957637   2.75  21.25  11.867143   
1961-11-05   5.88  15.79  11.571429  3.369201   3.96  13.46   9.590000   
1961-11-12   5.46  16.08  10.428571  3.939811   5.50  10.75   7.690000   
1961-11-19   7.50  15.00  10.798571  2.784358   4.21  13.00   7.951429   
1961-11-26   5.75  14.29   9.154286  3.214368   3.29  11.79   7.337143   
1961-12-03   7.92  23.75  12.608571  5.704669   4.67  18.71  10.442857   
1961-12-10  10.83  23.71  17.362857  4.890152   9.21  21.37  14.362857   
1961-12-17   9.29  21.34  14.985714  4.095106   8.63  17.62  13.472857   
1961-12-24  11.46  24.41  15.757143  4.959717   9.79  16.13  11.971429   
1961-12-31   5.04 

## The below exercises (67-70) required to use appl_1980_2014.csv  file

### 67. Import the file appl_1980_2014.csv and assign it to a variable called 'apple'

In [111]:
apple = pd.read_csv('Exercise Data Files/appl_1980_2014.csv')
apple

Date   Open   High    Low  Close     Volume  Adj Close
0     2014-07-08  96.27  96.80  93.92  95.35   65130000      95.35
1     2014-07-07  94.14  95.99  94.10  95.97   56305400      95.97
2     2014-07-03  93.67  94.10  93.20  94.03   22891800      94.03
3     2014-07-02  93.87  94.06  93.09  93.48   28420900      93.48
4     2014-07-01  93.52  94.07  93.13  93.52   38170200      93.52
...          ...    ...    ...    ...    ...        ...        ...
8460  1980-12-18  26.63  26.75  26.63  26.63   18362400       0.41
8461  1980-12-17  25.87  26.00  25.87  25.87   21610400       0.40
8462  1980-12-16  25.37  25.37  25.25  25.25   26432000       0.39
8463  1980-12-15  27.38  27.38  27.25  27.25   43971200       0.42
8464  1980-12-12  28.75  28.87  28.75  28.75  117258400       0.45

[8465 rows x 7 columns]

### 68.  Check out the type of the columns

In [112]:
apple.dtypes

Date          object
Open         float64
High         float64
Low          float64
Close        float64
Volume         int64
Adj Close    float64
dtype: object

### 69. Transform the Date column as a datetime type

In [113]:
apple['Date'] = pd.to_datetime(apple['Date'])
apple.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8465 entries, 0 to 8464
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   Date       8465 non-null   datetime64[ns]
 1   Open       8465 non-null   float64       
 2   High       8465 non-null   float64       
 3   Low        8465 non-null   float64       
 4   Close      8465 non-null   float64       
 5   Volume     8465 non-null   int64         
 6   Adj Close  8465 non-null   float64       
dtypes: datetime64[ns](1), float64(5), int64(1)
memory usage: 463.1 KB


### 70.  Set the date as the index

In [114]:
apple = apple.set_index('Date')
apple

Open   High    Low  Close     Volume  Adj Close
Date                                                        
2014-07-08  96.27  96.80  93.92  95.35   65130000      95.35
2014-07-07  94.14  95.99  94.10  95.97   56305400      95.97
2014-07-03  93.67  94.10  93.20  94.03   22891800      94.03
2014-07-02  93.87  94.06  93.09  93.48   28420900      93.48
2014-07-01  93.52  94.07  93.13  93.52   38170200      93.52
...           ...    ...    ...    ...        ...        ...
1980-12-18  26.63  26.75  26.63  26.63   18362400       0.41
1980-12-17  25.87  26.00  25.87  25.87   21610400       0.40
1980-12-16  25.37  25.37  25.25  25.25   26432000       0.39
1980-12-15  27.38  27.38  27.25  27.25   43971200       0.42
1980-12-12  28.75  28.87  28.75  28.75  117258400       0.45

[8465 rows x 6 columns]

### 71.  Is there any duplicate dates?

In [124]:
apple.index.duplicated().any()

False

### 72.  The index is from the most recent date. Sort the data so that the first entry is the oldest date.

In [127]:
apple = apple.sort_values(by = 'Date', axis = 0)
apple

Open   High    Low  Close     Volume  Adj Close
Date                                                        
1980-12-12  28.75  28.87  28.75  28.75  117258400       0.45
1980-12-15  27.38  27.38  27.25  27.25   43971200       0.42
1980-12-16  25.37  25.37  25.25  25.25   26432000       0.39
1980-12-17  25.87  26.00  25.87  25.87   21610400       0.40
1980-12-18  26.63  26.75  26.63  26.63   18362400       0.41
...           ...    ...    ...    ...        ...        ...
2014-07-01  93.52  94.07  93.13  93.52   38170200      93.52
2014-07-02  93.87  94.06  93.09  93.48   28420900      93.48
2014-07-03  93.67  94.10  93.20  94.03   22891800      94.03
2014-07-07  94.14  95.99  94.10  95.97   56305400      95.97
2014-07-08  96.27  96.80  93.92  95.35   65130000      95.35

[8465 rows x 6 columns]

### 73. Get the last business day of each month

In [140]:
apple['year_month'] = apple.index.strftime('%Y-%m')
apple.reset_index().groupby('year_month').agg(last_business_day = ('Date', 'max'))

last_business_day
year_month                  
1980-12           1980-12-31
1981-01           1981-01-30
1981-02           1981-02-27
1981-03           1981-03-31
1981-04           1981-04-30
...                      ...
2014-03           2014-03-31
2014-04           2014-04-30
2014-05           2014-05-30
2014-06           2014-06-30
2014-07           2014-07-08

[404 rows x 1 columns]

### 74.  What is the difference in days between the first day and the oldest

In [142]:
(apple.index.max() - apple.index.min()).days

12261

### 75.  How many months in the data we have?

In [143]:
(apple.index.max() - apple.index.min()).months

AttributeError: 'Timedelta' object has no attribute 'months'

## Section-8: The purpose of the below exercises is to understand how to create basic graphs

### 76. Plot the 'Adj Close' value. Set the size of the figure to 13.5 x 9 inches

## The below exercises (77-80) required to use Online_Retail.csv file

### 77. Import the dataset from this Online_Retail.csv and assign it to a variable called online_rt

### 78. Create a barchart with the 10 countries that have the most 'Quantity' ordered except UK

### 79.  Exclude negative Quatity entries

### 80. Create a scatterplot with the Quantity per UnitPrice by CustomerID for the top 3 Countries
Hint: First we need to find top-3 countries based on revenue, then create scater plot between Quantity and Unitprice for each country separately


## The below exercises (81-90) required to use FMCG_Company_Data_2019.csv file

### 81. Import the dataset FMCG_Company_Data_2019.csv and assign it to a variable called company_data

### 82. Create line chart for Total Revenue of all months with following properties
- X label name = Month
- Y label name = Total Revenue

### 83. Create line chart for Total Units of all months with following properties
- X label name = Month
- Y label name = Total Units
- Line Style dotted and Line-color should be red
- Show legend at the lower right location.

### 84. Read all product sales data (Facecream, FaceWash, Toothpaste, Soap, Shampo, Moisturizer) and show it  using a multiline plot
- Display the number of units sold per month for each product using multiline plots. (i.e., Separate Plotline for each product ).

### 85. Create Bar Chart for soap of all months and Save the chart in folder

### 86. Create Stacked Bar Chart for Soap, Shampo, ToothPaste for each month
The bar chart should display the number of units sold per month for each product. Add a separate bar for each product in the same chart.

### 87. Create Histogram for Total Revenue

### 88. Calculate total sales data (quantity) for 2019 for each product and show it using a Pie chart. Understand percentage contribution from each product

### 89. Create line plots for Soap & Facewash of all months in a single plot using Subplot

### 90. Create Box Plot for Total Profit variable